src/mpid/ch4/generic/mpidig_send.h

```diff
@@ -13,92 +13,127 @@
 #define MPIDIG_SEND_H_INCLUDED
 
 #include "ch4_impl.h"
+int MPIDIG_am_send_init(void);
+int MPIDIG_do_isend(const void *buf, MPI_Aint count, MPI_Datatype datatype,
+                    int rank, int tag, MPIR_Comm * comm, int context_offset,
+                    MPIDI_av_entry_t * addr, MPIR_Request * sreq, MPIR_Errflag_t errflag);
+
+int MPIDIG_am_ssend_init(void);
+int MPIDIG_do_ssend(const void *buf, MPI_Aint count, MPI_Datatype datatype,
+                    int rank, int tag, MPIR_Comm * comm, int context_offset,
+                    MPIDI_av_entry_t * addr, MPIR_Request * sreq, MPIR_Errflag_t errflag);
+
+int MPIDIG_am_long_init(void);
+int MPIDIG_do_long(const void *buf, MPI_Aint count, MPI_Datatype datatype,
+                   int rank, int tag, MPIR_Comm * comm, int context_offset,
+                   MPIDI_av_entry_t * addr, MPIR_Request * sreq, MPIR_Errflag_t errflag);
 
 #define MPIDI_NONBLOCKING 0
 #define MPIDI_BLOCKING    1
 
+#define MPIDIGI_AM_SEND_HDR_SIZE 16
+static inline int get_eager_limit(int is_local)
+{
+#ifdef MPIDI_CH4_DIRECT_NETMOD
+    return MPIDI_NM_am_hdr_max_sz() - MPIDIGI_AM_SEND_HDR_SIZE;
+#else
+    if (is_local) {
+        return MPIDI_SHM_am_hdr_max_sz() - MPIDIGI_AM_SEND_HDR_SIZE;
+    } else {
+        return MPIDI_NM_am_hdr_max_sz() - MPIDIGI_AM_SEND_HDR_SIZE;
+    }
+#endif
+}
+
```

Comment:
@raffenet Here is my approach to make eager/rndz decision at generic layer. Ideally, the am logic layer do not differentiate transport.